cpp/src/arrow/compute/kernels/scalar_temporal.cc

```diff
@@ -0,0 +1,509 @@
+};
+
+// ----------------------------------------------------------------------
+// Extract day of year from timestamp
+
+template <typename Duration>
+struct DayOfYear {
+  template <typename T, typename Arg>
+  static T Call(KernelContext*, Arg arg, Status*) {
+    const auto sd = sys_days{floor<days>(Duration{arg})};
+    return static_cast<T>((sd - sys_days(year_month_day(sd).year() / jan / 0)).count());
+  }
+};
+
+// ----------------------------------------------------------------------
+// Extract ISO Year values from timestamp
+
+template <typename Duration>
+struct ISOYear {
+  template <typename T, typename Arg>
+  static T Call(KernelContext*, Arg arg, Status*) {
+    return static_cast<T>(static_cast<const int32_t>(
+        year_month_day{sys_days{floor<days>(Duration{arg})} + days{3}}.year()));
+  }
+};
+
+// ----------------------------------------------------------------------
+// Extract ISO week from timestamp
+
+// Based on
+// https://github.com/HowardHinnant/date/blob/6e921e1b1d21e84a5c82416ba7ecd98e33a436d0/include/date/iso_week.h#L1503
+template <typename Duration>
+struct ISOWeek {
+  template <typename T, typename Arg>
+  static T Call(KernelContext*, Arg arg, Status*) {
+    const auto dp = sys_days{floor<days>(Duration{arg})};
+    auto y = year_month_day{dp + days{3}}.year();
+    auto start = sys_days((y - years{1}) / dec / thu[last]) + (mon - thu);
+    if (dp < start) {
+      --y;
+      start = sys_days((y - years{1}) / dec / thu[last]) + (mon - thu);
+    }
+    return static_cast<T>(trunc<weeks>(dp - start).count() + 1);
+  }
+};
+
+// ----------------------------------------------------------------------
+// Extract ISO calendar values from timestamp
+
+const std::shared_ptr<DataType> iso_calendar_type = struct_(
+    {field("iso_year", int64()), field("iso_week", int64()), field("weekday", int64())});
+
+template <typename Duration>
+struct ISOCalendar {
+  static Status Call(KernelContext* ctx, const Scalar& in, Scalar* out) {
+    using ScalarType = typename TypeTraits<Int64Type>::ScalarType;
+
+    const auto& in_val = internal::UnboxScalar<const TimestampType>::Unbox(in);
+    const auto dp = sys_days{floor<days>(Duration{in_val})};
+    const auto ymd = year_month_day(dp);
+    auto y = year_month_day{dp + days{3}}.year();
+    auto start = sys_days((y - years{1}) / dec / thu[last]) + (mon - thu);
+    if (dp < start) {
+      --y;
+      start = sys_days((y - years{1}) / dec / thu[last]) + (mon - thu);
+    }
+
+    std::vector<std::shared_ptr<Scalar>> values = {
+        std::make_shared<ScalarType>(
+            static_cast<int64_t>(static_cast<int32_t>(ymd.year()))),
+        std::make_shared<ScalarType>(
+            static_cast<int64_t>(trunc<weeks>(dp - start).count() + 1)),
+        std::make_shared<ScalarType>(static_cast<int64_t>(weekday(ymd).iso_encoding()))};
+    *checked_cast<StructScalar*>(out) =
+        StructScalar(std::move(values), iso_calendar_type);
+    return Status::OK();
+  }
```

Comment:
@pitrou any idea why this scalar kernel is failing here? The test outputs and [CI log](https://github.com/apache/arrow/pull/10176/checks?check_run_id=2666780178#step:10:1299) is not very telling.

Reply:
Got it. Kernel signature was set to only expect arrays and not scalars.